applications/StructuralMechanicsApplication/custom_elements/timoshenko_beam_element_2D2N.cpp

```diff
@@ -0,0 +1,1051 @@
+            mConstitutiveLawVector[point_number] = GetProperties()[CONSTITUTIVE_LAW]->Clone();
+            mConstitutiveLawVector[point_number]->InitializeMaterial(r_properties, r_geometry, N_values);
+        }
+    } else
+        KRATOS_ERROR << "A constitutive law needs to be specified for the element with ID " << this->Id() << std::endl;
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+Element::Pointer TimoshenkoBeamElement2D2N::Clone(
+    IndexType NewId,
+    NodesArrayType const& rThisNodes
+    ) const
+{
+    KRATOS_TRY
+
+    TimoshenkoBeamElement2D2N::Pointer p_new_elem = Kratos::make_intrusive<TimoshenkoBeamElement2D2N>
+        (NewId, GetGeometry().Create(rThisNodes), pGetProperties());
+    p_new_elem->SetData(this->GetData());
+    p_new_elem->Set(Flags(*this));
+
+    // Currently selected integration methods
+    p_new_elem->SetIntegrationMethod(mThisIntegrationMethod);
+
+    // The vector containing the constitutive laws
+    p_new_elem->SetConstitutiveLawVector(mConstitutiveLawVector);
+
+    return p_new_elem;
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::EquationIdVector(
+    EquationIdVectorType& rResult,
+    const ProcessInfo& rCurrentProcessInfo
+    ) const
+{
+    KRATOS_TRY;
+
+    const auto& r_geom = GetGeometry();
+    const SizeType number_of_nodes = r_geom.size();
+    const SizeType dofs_per_node = GetDoFsPerNode(); // u, v, theta
+
+    if (rResult.size() != dofs_per_node * number_of_nodes)
+        rResult.resize(dofs_per_node * number_of_nodes);
+
+
+    for (IndexType i = 0; i < number_of_nodes; ++i) {
+        const SizeType index = i * dofs_per_node;
+        rResult[index]     = r_geom[i].GetDof(DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = r_geom[i].GetDof(DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = r_geom[i].GetDof(ROTATION_Z    ).EquationId();
+    }
+
+    KRATOS_CATCH("")
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::GetDofList(
+    DofsVectorType& rElementalDofList,
+    const ProcessInfo& rCurrentProcessInfo
+    ) const
+{
+    KRATOS_TRY;
+
+    const auto& r_geom = GetGeometry();
+    const SizeType number_of_nodes = r_geom.size();
+    const SizeType dofs_per_node = GetDoFsPerNode(); // u, v, theta
+    rElementalDofList.resize(dofs_per_node * number_of_nodes);
+
+    for (IndexType i = 0; i < number_of_nodes; ++i) {
+        const SizeType index = i * dofs_per_node;
+        rElementalDofList[index]     = r_geom[i].pGetDof(DISPLACEMENT_X);
```

Comment:
you could use the GetDof by key, it is a bit faster (see the fluids elements )

Reply:
DONE